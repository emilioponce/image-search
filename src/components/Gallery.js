import React, { Component } from 'react'
import Image from './Image'

const Gallery = () => {
  return (
    <div>
      <Image />
      <Image />
      <Image />
    </div>
  )
}

export default Gallery
