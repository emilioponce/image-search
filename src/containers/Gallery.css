.Gallery {
  border-top: 1px solid #202329;
  margin-bottom: 30px;
  background-color: #323841;
}

@supports (display: grid) {
  .Images {
    display: grid;
    grid-template-columns: repeat(auto-fit, 150px);
    grid-gap: 2rem;
  }

  .Images {
    margin: 0;
  }
}

.pagination {
  display: inline-block;
  padding-left: 0;
  padding-top: 10px;
}

.pagination > li {
  display: inline;
  padding: 6px 16px;
  background-color: #202329;
  margin: 1px;
}

.pagination > li.active {
  background-color: #61dafb;
  color: #000000;
}

.pagination > li.active > a {
  color: #000000;
  text-decoration: none;
}

.pagination > li:not(.active) > a {
  color: #ffffff;
  text-decoration: none;
}

.pagination > li:hover:not(.active) {
  background-color: black;
  color: white;
}

.Loading {
  min-height: 500px;
  display: flex;
  align-items: center;
  justify-content: center;
}

.Counter {
  display: inline-block;
  font-size: 80%;
  color: white;
}
