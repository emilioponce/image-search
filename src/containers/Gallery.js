import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import _ from 'lodash'
import PropTypes from 'prop-types'
import Pagination from 'react-js-pagination'
import Loader from 'react-loader-spinner'

import Image from './Image'
import { fetchImages } from '../actions'
import { PAGE_RANGE } from '../config/constants'
import { getSearchResultsSelector } from '../reducers/search'

import './Gallery.css'

const Gallery = ({ images, keyword, page, loading, fetchImages }) => {
  const handlePageChange = page => {
    fetchImages(keyword, page)
  }

  if (loading) {
    return (
      <div className="Gallery">
        <div className="Loading">
          <Loader type="TailSpin" color="#61dafb" height="100" width="100" />
        </div>
      </div>
    )
  }

  if (_.isEmpty(images)) {
    return <div className="Gallery" />
  }

  let photos = images.photo

  if (_.isEmpty(photos)) {
    return <div className="Gallery">No results found</div>
  }

  return (
    <div className="Gallery">
      <Pagination
        activePage={page}
        itemsCountPerPage={images.perpage}
        totalItemsCount={images.total}
        pageRangeDisplayed={PAGE_RANGE}
        onChange={handlePageChange}
      />
      <div className="Images">
        {photos.map(photo => (
          <Image
            key={photo.id}
            url={photo.url_q}
            urlBig={photo.url_m}
            title={photo.title}
            date={photo.dateupload}
            owner={photo.ownername}
            description={photo.description}
          />
        ))}
      </div>
      <Pagination
        activePage={page}
        itemsCountPerPage={images.perpage}
        totalItemsCount={images.total}
        pageRangeDisplayed={PAGE_RANGE}
        onChange={handlePageChange}
      />
    </div>
  )
}

const mapStateToProps = state => {
  return getSearchResultsSelector(state)
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ fetchImages }, dispatch)
  /* NOTE that exists an alternative way without bindActionCreators. Dispatching 
  the action by hand. IT's more explicit, but verbose.
   return {
     fetchImages: (keyword, page) => dispatch(fetchImages(keyword, page))
   } 
   */
}

Gallery.propTypes = {
  images: PropTypes.object.isRequired,
  keyword: PropTypes.string.isRequired,
  page: PropTypes.number.isRequired,
  loading: PropTypes.bool.isRequired,
  fetchImages: PropTypes.func.isRequired
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Gallery)
