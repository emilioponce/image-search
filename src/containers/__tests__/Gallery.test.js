import React from 'react'
import ReactDOM from 'react-dom'
import { shallow } from 'enzyme'
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { Gallery } from '../Gallery'

configure({ adapter: new Adapter() })

const mockFetchImagesFn = jest.fn()

const component = (
  <Gallery
    images={{}}
    keyword={'cats'}
    page={1}
    loading={false}
    fetchImages={mockFetchImagesFn}
  />
)

describe('Gallery component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(component)
  })

  it('Renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(component, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('Snapshot matches', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it("Should show a Spinner if it's loading", () => {
    const component = (
      <Gallery
        images={{}}
        keyword={'cats'}
        page={1}
        loading={true}
        fetchImages={mockFetchImagesFn}
      />
    )
    let wrapper = shallow(component)
    expect(wrapper.find('.Gallery').text()).toBe('<Loader />')
  })

  it('Should show an empty gallery', () => {
    expect(wrapper.find('.Gallery').text()).toBe('')
  })

  it('Should show a message for no results found', () => {
    const component = (
      <Gallery
        images={{ photos: {} }}
        keyword={'cats'}
        page={1}
        loading={false}
        fetchImages={mockFetchImagesFn}
      />
    )
    let wrapper = shallow(component)
    expect(wrapper.find('.Gallery').text()).toBe('No results found')
  })

  // NOTE:  we don't test Pagination as is a 3rd party library
})
