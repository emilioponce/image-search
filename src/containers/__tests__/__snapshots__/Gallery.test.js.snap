// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery component Snapshot matches 1`] = `
<div
  className="Gallery"
/>
`;
